### DISCUSSION TITLE
Question - 802.1Qbu frame preemption implementation

### DISCUSSION THREAD

**Discussion by yue2388253** [CoRE-RG/CoRE4INET#30] [Ideas]
Hi,

I was wondering whether CoRE4INET support IEEE 802.1Qbu frame preemption?

Best
Lam

**Comment by Musteblume** [comment]
Hi Lam

no preemption is not supported by the public CoRE4INET version.
We have some old implementation on a different branch, but this is not compatible with the newer contained Qbv shaper implementation.

Best regards
Philipp

**Comment by yue2388253** [comment]
Hi Philipp,

Thank you for reply.
Are there any plans to implement this feature?
Another question is that whether a lookahead mechanism is enabled in switches. Lookahead mechanism means that in 802.1Qbv network, the switch will not transmit a frame if the remaining open time is too short to transmit the frame.

Best,
Lam

**Comment by Musteblume** [comment]
Hi Lam,

currently there are no plans for preemption.
The next inet-framework version will include many TSN features in addition to preemption: [https://www.youtube.com/watch?v=5kn7eQkvdUg](url)


The lookahead mechanism is currently not implemented in the Qbv shaper.
You could extend IEEE8021QbvGate::isOpen() to look for the size of the next frame in the queue and compare it with the next timestamp in the Gate Control List.

Hope this helps
Best regards
Philipp

**Comment by yue2388253** [comment]
Hi Philipp,

Thank you for your reply! 
I have tried to use INET 4.3.0 but it seems the TSN-related features are not fully implemented and it has limited examples about TSN, which make it difficult to use.
And I also tried NeSTing but it fails to run CQF since 802.1Qci is not implemented.
Will CoRE4INET continue to update TSN-related features? (For example, 802.1Qcr, parternoster queuing and so on)

Best, 
Lam

**Comment by Musteblume** [comment]
Hi Lam,

INET 4.4.x will include many TSN-related features with many examples and documentation. (Qcr is also included)

We actively use CoRE4INET for our research currently and in the future.
So, we implement features that we need for our research and publish all mature extensions. 
Thats why i cannot say which features will be added in the future.
We are also happy to integrate pull requests with features implemented by others.

Hope this answers your question
Best regards
Philipp
